### ISSUE TITLE
Support controlling thread affinity

### ISSUE THREAD

**Issue by tmds** [dotnet/runtime#21363]
**Purpose**

Allow control of thread affinity.

**Proposed API**

```C#
class Thread
{
    ProcessorSet Affinity { get; set; }
}
```
The `ProcessorSet` represents a a set of logical processors.
The underlying APIs for Windows/Posix use bit masks for this.

Windows: https://msdn.microsoft.com/en-us/library/windows/desktop/ms686247(v=vs.85).aspx
Posix: http://man7.org/linux/man-pages/man3/pthread_setaffinity_np.3.html

On Windows, systems with more than 64 processors are split into groups. Threads cannot use processors which belong to multiple groups. `ProcessorSet` can store the group info.

```C#
struct ProcessorSet
{
   int Group { get; set; }                  // get/set group
   bool this[int processor] { get; set; }   // get/set/clear processor

   ProcessorSet(minCapacity);               // actual capacity may be higher
   int Capacity { get; }                    // indexer is valid from [0-Capacity[

   // string representation, see below
   string ToString();
   static ProcessorSet Parse(string);
   static bool TryParse(string, out ProcessorSet);

   static ProcessorSet ForProcessor(int group, int processor); // set with single processor
   static ProcessorSet ForProcessor(int processor);            // set with single processor
}
```

The string representation can be a comma separated list of processors and ranges, prefixed with an optional processor group. E.g: `1`, `1-3,5`, `1:8-10`.

**Example**

```
var thread = new Thread(threadStart);
thread.Affinity = ProcessorSet.Parse("0-3");
thread.Start();
```

**Remarks**

This proposal does not consider retrieving information from the OS like the number of groups and the number of processors per group. The total number of processors can be read from `Environment.ProcessorCount`.

Posix threads inherit the calling thread affinity. It may be preferable to allow new Threads to run on all processors to avoid surprises. Threads that want to inherit the calling thread priority can explicitly do so:

```c#
newThread.Affinity = Thread.CurrentThread.Affinity;
```

**Comment by john-h-k** [comment]
> Even without support for ProcessorSets, .Net should support `Thread.ProcessorAffinity` as a bit mask, similar to Process.ProcessorAffinity. There is no alternative cross-platform API besides rolling our own based on `sched_setaffinity` and `SetThreadAffinityMask`.

Yep, which fully verges on undefined behaviour bc we don't know if `Thread` is backed by a single native OS thread

**Comment by KalleOlaviNiemitalo** [comment]
> we don't know if `Thread` is backed by a single native OS thread

There is [Thread.BeginThreadAffinity](https://docs.microsoft.com/dotnet/api/system.threading.thread.beginthreadaffinity?view=net-5.0) for asking the host not to move the managed thread to a different OS thread. <https://github.com/dotnet/runtime/blob/2f740adc1457e8a28c1c072993b66f515977eb51/src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs#L209-L214>

**Comment by tannergooding** [comment]
That isn't quite the same as enforcing a thread always executes on a given CPU core, which is the intent for the proposed API.

**Comment by KalleOlaviNiemitalo** [comment]
I meant, if an application is going to call OS functions to set the CPU affinity of the OS thread, it ought to call Thread.BeginThreadAffinity first. The implementation seems to be a no-op in .NET 5 but the documentation does not promise that.

If an API for setting the CPU affinity of a managed thread is added and the application calls that, I imagine the runtime will have to avoid running the managed thread on an OS thread that has a different CPU affinity, and the application won't need to call Thread.BeginThreadAffinity for that purpose.

**Comment by ddrinka** [comment]
Our .Net Framework code utilizes `BeginThreadAffinity` to bind a .Net thread to an OS thread, and then iterates through `Process.GetCurrentProcess().Threads` until finding the `ProcessThread` who's ID matches `AppDomain.GetCurrentThreadId()`.  We can then manage the affinity of that ProcessThread.

Our modern code uses the following PInvokes, which also work fine:

```csharp
public static void SetThreadProcessorAffinity(params int[] processors)
{
    Thread.BeginThreadAffinity();

    var mask = ComputeProcessorMask(processors);

    if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
    {
        sched_setaffinity(0, new IntPtr(sizeof(ulong)), new[] { mask });
    }
    else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
    {
        var currentThread = GetCurrentThread();
        SetThreadAffinityMask(currentThread, mask);
    }
    else
    {
        throw new NotSupportedException();
    }
}
```

However, we believe this capability should be included in the runtime.

**Comment by quixoticaxis** [comment]
Any resolution on this issue? I've followed this and linked issues, and there seems to be no decision whether this should be enabled or not.

Disclaimer: bumping this issue to prevent automatic cleanup.
